macOS/DuckDuckGo/Tab/TabExtensions/TabCrashRecoveryExtension.swift

```diff
@@ -0,0 +1,145 @@
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import BrowserServicesKit
+import Combine
+import Foundation
+import Navigation
+import WebKit
+import PixelKit
+
+/**
+ * This Tab Extension is responsible for recovering from tab crashes.
+ */
+final class TabCrashRecoveryExtension {
+    private weak var webView: WKWebView?
+    private var content: Tab.TabContent?
+    private var lastCrashedAt: Date?
+    private let featureFlagger: FeatureFlagger
+    private var webViewError: WKError?
+    private let tabCrashErrorSubject = PassthroughSubject<TabCrashErrorPayload, Never>()
+    private let firePixel: (PixelKitEvent, [String: String]) -> Void
+
+    private var cancellables = Set<AnyCancellable>()
+
+    enum Const {
+        static let crashLoopInterval: TimeInterval = 5
+    }
+
+    init(
+        featureFlagger: FeatureFlagger,
+        contentPublisher: some Publisher<Tab.TabContent, Never>,
+        webViewPublisher: some Publisher<WKWebView, Never>,
+        webViewErrorPublisher: some Publisher<WKError?, Never>,
+        firePixel: @escaping (PixelKitEvent, [String: String]) -> Void = { event, parameters in
+            PixelKit.fire(event, frequency: .dailyAndStandard, withAdditionalParameters: parameters)
+        }
+    ) {
+        self.featureFlagger = featureFlagger
+        self.firePixel = firePixel
+
+        contentPublisher.sink { [weak self] content in
+            self?.content = content
+        }
+        .store(in: &cancellables)
+
+        webViewErrorPublisher.sink { [weak self] webViewError in
+            self?.webViewError = webViewError
+        }
+        .store(in: &cancellables)
+
+        webViewPublisher.sink { [weak self] webView in
+            self?.webView = webView
+        }.store(in: &cancellables)
+    }
+}
+
+extension TabCrashRecoveryExtension: NavigationResponder {
+
+    func webContentProcessDidTerminate(with reason: WKProcessTerminationReason?) {
+        guard let webView, (webViewError?.code.rawValue ?? WKError.Code.unknown.rawValue) != WKError.Code.webContentProcessTerminated.rawValue else {
+            return
+        }
+
+        let terminationReason = reason?.rawValue ?? -1
+
+        let error = WKError(.webContentProcessTerminated, userInfo: [
+            WKProcessTerminationReason.userInfoKey: terminationReason,
+            NSLocalizedDescriptionKey: UserText.webProcessCrashPageMessage,
+            NSUnderlyingErrorKey: NSError(domain: WKErrorDomain, code: terminationReason)
+        ])
+
+        Task {
+#if APPSTORE
+            let additionalParameters = [String: String]()
+#else
+            let additionalParameters = await SystemInfo.pixelParameters()
+#endif
+            firePixel(DebugEvent(GeneralPixel.webKitDidTerminate, error: error), additionalParameters)
+        }
+
+        if featureFlagger.isFeatureOn(.tabCrashRecovery) {
```

Comment:
@ayoy What are your thought on making all this logic more decoupled so we can add unit tests to it?